unix: reopen tty as /dev/tty

saghul opened the pull request:
Reopen the file descriptor when it refers to a tty. This lets us put the tty in
non-blocking mode without affecting other processes that share it with
us.

This brings back commit 31f9fbc, which was reverted in 20bb1bf. The OSX
select trick is working now.

Original patch by @bnoordhuis

saghul committed:
unix: reopen tty as /dev/tty

Reopen the file descriptor when it refers to a tty. This lets us put the
tty in non-blocking mode without affecting other processes that share it
with us.

This brings back commit 31f9fbc, which was reverted in 20bb1bf. The OSX
select trick is working now.

Original patch by @bnoordhuis

Nodejs-Jenkins commented:
Thank you for contributing this pull request! Here are a few pointers to make sure your submission will be considered for inclusion.

Commit saghul/libuv@e22057c89bddd904fa51d00a8de086e40dbfa593 has the following error(s):
- Commit message line too long: 2

The following commiters were not found in the CLA:
- Saúl Ibarra Corretgé

You can fix all these things _without_ opening another issue.

Please see [CONTRIBUTING.md](https://github.com/joyent/node/blob/master/CONTRIBUTING.md) for more information

saghul commented:
So, I un-reverted and adapted the un-revert, tried the failing case `node | cat` and ran the test suite and it seems to work!

/cc @indutny, @bnoordhuis: can you guys review?

PS: I still need to look into making the newdf close-on-exec.

indutny commented:
Needs close-on-exec, otherwise harmful :)

saghul commented:
Force pushed!

bnoordhuis commented on src/unix/tty.c:
```diff
@@ -35,26 +35,70 @@
+   * other processes.
+   */
+  if (isatty(fd)) {
+    newfd = open("/dev/tty", O_RDWR);
```
There is a `uv__open_cloexec()` in src/unix/stream.c that you can use.

bnoordhuis commented on src/unix/tty.c:
```diff
@@ -35,26 +35,70 @@
+      return -errno;
+
+    do
+      r = dup2(newfd, fd);
```
This should use dup3() or fcntl(F_DUP2FD_CLOEXEC) where possible.

bnoordhuis replied:
And use uv__cloexec() when those are unavailable, of course.

saghul commented:
@bnoordhuis I addressed your feedback, can you check again? I added a `uv__dup_cloexec` helper funcion, should we ever need it. It uses:
- dup3 on Linux (if possible)
- dup3 on FreeBSD 10
- fcntl with F_DUP2FD_CLOEXEC (if possible)
- dup2 + uv__cloexec as a last resort

Hope I got it right :-)

saghul commented:
Oh, I'll rename uv__dup_cloexec to uv__dup2_cloexec, kinda makes more sense.

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -93,6 +93,52 @@ static int uv__open_cloexec(const char* path, int flags) {
+      no_dup3 = 1;
+    else if (r == -1)
+      return -errno;
+    return 0;
```
Should return r.

saghul replied:
I thought of that, but what should the fallback case (dup2 + cloexec)
return?
On Mar 8, 2014 4:52 PM, "Ben Noordhuis" notifications@github.com wrote:

> In src/unix/stream.c:
> 
> > @@ -93,6 +93,52 @@ static int uv__open_cloexec(const char\* path, int flags) {
> >  }
> > 
> > +int uv__dup_cloexec(int oldfd, int newfd) {
> > -  int r;
> >   +
> >   +#if defined(**linux**)
> > -  static int no_dup3;
> > -  if (!no_dup3) {
> > -    do
> > -      r = uv__dup3(oldfd, newfd, UV__O_CLOEXEC);
> > -    while (r == -1 && (errno == EINTR || errno == EBUSY));
> > -    if (r == -1 && errno == ENOSYS)
> > -      no_dup3 = 1;
> > -    else if (r == -1)
> > -      return -errno;
> > -    return 0;
> 
> Should return r.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/joyent/libuv/pull/1184/files#r10409652
> .

bnoordhuis replied:
In the inconceivable case that uv__cloexec() fails, the return value of uv__cloexec() (and it should close the file descriptor).  Else it should return the new file descriptor.

saghul replied:
Sounds reasonable, thanks a lot for the help Ben.
On Mar 8, 2014 4:57 PM, "Ben Noordhuis" notifications@github.com wrote:

> In src/unix/stream.c:
> 
> > @@ -93,6 +93,52 @@ static int uv__open_cloexec(const char\* path, int flags) {
> >  }
> > 
> > +int uv__dup_cloexec(int oldfd, int newfd) {
> > -  int r;
> >   +
> >   +#if defined(**linux**)
> > -  static int no_dup3;
> > -  if (!no_dup3) {
> > -    do
> > -      r = uv__dup3(oldfd, newfd, UV__O_CLOEXEC);
> > -    while (r == -1 && (errno == EINTR || errno == EBUSY));
> > -    if (r == -1 && errno == ENOSYS)
> > -      no_dup3 = 1;
> > -    else if (r == -1)
> > -      return -errno;
> > -    return 0;
> 
> In the inconceivable case that uv__cloexec() fails, the return value of
> uv__cloexec() (and it should close the file descriptor). Else it should
> return the new file descriptor.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/joyent/libuv/pull/1184/files#r10409677
> .

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -93,6 +93,52 @@ static int uv__open_cloexec(const char* path, int flags) {
+  while (r == -1 && errno == EINTR);
+  if (r == -1)
+    return -errno;
+  return 0;
```
Ditto.

bnoordhuis commented on src/unix/tty.c:
```diff
@@ -35,26 +35,61 @@
+      return -errno;
+
+    r = uv__dup_cloexec(newfd, fd);
+    if (r && r != -EINVAL) {
```
When you update uv__dup_cloexec(), you need to update this check as well.

bnoordhuis commented:
Maybe move uv__dup_cloexec() and uv__open_cloexec() to src/unix/core.c, that's where most of the system call polyfills live.

saghul commented:
@bnoordhuis addressed your last comments.

bnoordhuis commented on src/unix/core.c:
```diff
@@ -821,3 +821,87 @@ int uv_getrusage(uv_rusage_t* rusage) {
+  if (errno != EINVAL)
+    return -errno;
+
+  /* O_CLOEXEC not supported. */
```
It would be good to cache the result in a no_cloexec static variable so you don't keep repeating the open(O_CLOEXEC) call when it's unsupported.

bnoordhuis commented on src/unix/core.c:
```diff
@@ -821,3 +821,87 @@ int uv_getrusage(uv_rusage_t* rusage) {
+      no_dup3 = 1;
+    else if (r == -1)
+      return -errno;
+    return r;
```
Returns -1 when the system call failed with ENOSYS.  Probably not what you intended, I suspect you want it to fall through.

bnoordhuis commented on src/unix/core.c:
```diff
@@ -821,3 +821,87 @@ int uv_getrusage(uv_rusage_t* rusage) {
+  while (r == -1 && errno == EINTR);
+  if (r == -1)
+    return -errno;
+  return r;
```
You may want to consider falling through on EINVAL in case libuv is compiled against kernel headers that are newer than the actual kernel.

bnoordhuis commented on src/unix/core.c:
```diff
@@ -821,3 +821,87 @@ int uv_getrusage(uv_rusage_t* rusage) {
+    int err;
+    do
+      r = dup2(oldfd, newfd);
+    while (r == -1 && (errno == EINTR || errno == EBUSY));
```
Minor caveat: EBUSY is a Linux-ism.  The BSDs never return it and I'm pretty sure Solaris doesn't either.

Probably doesn't hurt now but might be good to exclude it, just in case dup2() in the future does start returning EBUSY but with semantics that are different from Linux.

bnoordhuis commented on src/unix/linux-syscalls.c:
```diff
@@ -422,3 +434,12 @@ ssize_t uv__pwritev(int fd, const struct iovec *iov, int iovcnt, off_t offset) {
+
+int uv__dup3(int oldfd, int newfd, int flags) {
+#if defined(__NR_dup3)
+  return syscall(__NR_dup3, oldfd, newfd);
```
You forgot the flags argument.

saghul replied:
What the... :-/

bnoordhuis commented on src/unix/tty.c:
```diff
@@ -35,26 +35,61 @@
+    flags = UV_STREAM_WRITABLE;
+
+  uv__nonblock(fd, 1);
+  uv__stream_open((uv_stream_t*)tty, fd, flags);
```
Style is inconsistent with a few lines up: `(uv_stream_t*)tty` vs. `(uv_stream_t*) tty`

saghul commented:
Thanks for the detailed feeback @bnoordhuis! I think I now coevered all those blind spots :-)

bnoordhuis commented on src/unix/core.c:
```diff
@@ -821,3 +821,101 @@ int uv_getrusage(uv_rusage_t* rusage) {
+  while (r == -1 && errno == EINTR);
+  if (r == -1)
+    return -errno;
+  return r;
```
This will make the compiler complain about unreachable code and the 'fallback' label being unused.  I think you can rewrite the error checking logic in all three cases to something like this:

``` c
if (r != -1)
  return r;
if (errno != EINVAL)
  return -errno;
/* Fall through. */
```

saghul replied:
Changed. How do you suggest I initialize `err`? In case of FreeBSD 10 the fallthrough case will never be reached.

bnoordhuis replied:
Maybe like this?

``` c
int uv__dup2_cloexec(int oldfd, int newfd) {
#if defined(__FreeBSD__) && __FreeBSD__ >= 10
  int r;
  do
    r = dup3(oldfd, newfd, O_CLOEXEC);
  while (r == -1 && errno == EINTR);
  if (r < 0)
    return -errno;
  return r;
#else
# if defined(__linux__)
  // ...
# elif /* etc. */
  // ...
# endif
#endif
/* fall through goes here */
#endif
}
```

saghul replied:
Updated, how about now?

bnoordhuis commented on src/unix/core.c:
```diff
@@ -821,3 +823,93 @@ int uv_getrusage(uv_rusage_t* rusage) {
+  static int no_cloexec;
+
+  if (!no_cloexec) {
+    fd = open(path, flags | UV__O_CLOEXEC);
```
I believe FreeBSD 10 also supports open(O_CLOEXEC) but apart from that LGTM.

saghul replied:
On 03/10/2014 06:04 PM, Ben Noordhuis wrote:

> In src/unix/core.c:
> 
> > @@ -821,3 +823,93 @@ int uv_getrusage(uv_rusage_t\* rusage) {
> > 
> >    return 0;
> >  }
> > +
> > +
> > +int uv__open_cloexec(const char\* path, int flags) {
> > -  int err;
> > -  int fd;
> >   +
> >   +#if defined(**linux**)
> > -  static int no_cloexec;
> >   +
> > -  if (!no_cloexec) {
> > -    fd = open(path, flags | UV__O_CLOEXEC);
> 
> I believe FreeBSD 10 also supports open(O_CLOEXEC) but apart from that LGTM.

Updated. Online man pages say it's supported since version 9, but I 
could only tested in 10, where it does work.

bnoordhuis commented:
I'll take your word for it. ;-)  LGTM.

saghul commented:
Thanks for your invaluable help Ben, I really appreciate it. Landed!
